primitives/weights/src/weight_v2.rs

```diff
@@ -48,71 +51,118 @@ impl Weight {
 		self
 	}
 
+	/// Set the storage size part of the weight.
+	pub const fn set_proof_size(mut self, c: u64) -> Self {
+		self.proof_size = c;
+		self
+	}
+
 	/// Return the reference time part of the weight.
 	pub const fn ref_time(&self) -> u64 {
 		self.ref_time
 	}
 
-	/// Return a mutable reference time part of the weight.
+	/// Return the storage size part of the weight.
+	pub const fn proof_size(&self) -> u64 {
+		self.proof_size
+	}
+
+	/// Return a mutable reference to the reference time part of the weight.
 	pub fn ref_time_mut(&mut self) -> &mut u64 {
 		&mut self.ref_time
 	}
 
-	pub const MAX: Self = Self { ref_time: u64::MAX };
+	/// Return a mutable reference to the storage size part of the weight.
+	pub fn proof_size_mut(&mut self) -> &mut u64 {
+		&mut self.proof_size
+	}
+
+	pub const MAX: Self = Self { ref_time: u64::MAX, proof_size: u64::MAX };
 
 	/// Get the conservative min of `self` and `other` weight.
 	pub fn min(&self, other: Self) -> Self {
-		Self { ref_time: self.ref_time.min(other.ref_time) }
+		Self {
+			ref_time: self.ref_time.min(other.ref_time),
+			proof_size: self.proof_size.min(other.proof_size),
+		}
 	}
 
 	/// Get the aggressive max of `self` and `other` weight.
 	pub fn max(&self, other: Self) -> Self {
-		Self { ref_time: self.ref_time.max(other.ref_time) }
+		Self {
+			ref_time: self.ref_time.max(other.ref_time),
+			proof_size: self.proof_size.max(other.proof_size),
+		}
 	}
 
 	/// Try to add some `other` weight while upholding the `limit`.
 	pub fn try_add(&self, other: &Self, limit: &Self) -> Option<Self> {
 		let total = self.checked_add(other)?;
-		if total.ref_time > limit.ref_time {
+		if total.ref_time > limit.ref_time || total.proof_size > limit.proof_size {
 			None
 		} else {
 			Some(total)
 		}
 	}
 
-	/// Construct [`Weight`] with reference time weight.
+	/// Construct [`Weight`] with reference time weight and 0 storage size weight.
 	pub const fn from_ref_time(ref_time: u64) -> Self {
-		Self { ref_time }
+		Self { ref_time, proof_size: 0 }
 	}
 
+	/// Construct [`Weight`] with storage size weight and 0 reference time weight.
+	pub const fn from_proof_size(proof_size: u64) -> Self {
+		Self { ref_time: 0, proof_size }
+	}
+
+	/// Construct [`Weight`] with weight components, namely reference time and storage size weights.
+	pub const fn from_components(ref_time: u64, proof_size: u64) -> Self {
+		Self { ref_time, proof_size }
+	} 
+
 	/// Saturating [`Weight`] addition. Computes `self + rhs`, saturating at the numeric bounds of
 	/// all fields instead of overflowing.
 	pub const fn saturating_add(self, rhs: Self) -> Self {
-		Self { ref_time: self.ref_time.saturating_add(rhs.ref_time) }
+		Self {
+			ref_time: self.ref_time.saturating_add(rhs.ref_time),
+			proof_size: self.proof_size.saturating_add(rhs.proof_size),
+		}
```

Comment:
Just an idea: Could have opted for a private field iter, such that this becomes something like:

```suggestion
		self.fields_iter().map(|s| saturating_add(s, rhs)).collect::<Self>()
```

Then it is future proof when changing or adding fields. But could be overkill.

Reply:
That also assumes that all fields are `u64`, which might not be the case in the future.

Reply:
No... it just assumes that they are `Saturating`. This iter would probably not be a homogeneous collection but a custom type.

Reply:
That, and the fact that all the fields must be homogeneous, because you're mapping over a collection of `Item`s, which must have the same type.